File: apps/api-go/internal/domain/usecase/get_insights.go
package usecase

import (
	"context"

	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/port"
)

// defaultInsightFields are the metrics requested when none are specified.
var defaultInsightFields = []string{"impressions", "clicks", "spend", "ctr", "cpc", "reach"}

// GetInsightsUseCase fetches performance metrics for a campaign.
type GetInsightsUseCase struct {
	adPlatform port.AdPlatform
	repo       port.CampaignRepository
}

func NewGetInsightsUseCase(adPlatform port.AdPlatform, repo port.CampaignRepository) *GetInsightsUseCase {
	return &GetInsightsUseCase{adPlatform: adPlatform, repo: repo}
}

// Execute returns insights for the campaign over [since, until]. A campaign that
// was never pushed to Meta has no metrics, so we return an empty data set.
func (uc *GetInsightsUseCase) Execute(ctx context.Context, campaignID, since, until string) (map[string]any, error) {
	campaign, err := uc.repo.Get(ctx, campaignID)
	if err != nil {
		return nil, err
	}
	if campaign.MetaCampaignID == "" {
		return map[string]any{"data": []any{}}, nil
	}
	return uc.adPlatform.GetInsights(ctx, campaign.MetaCampaignID, defaultInsightFields, since, until)
}

File: apps/api/domain/use_cases/get_insights.py
"""Use case: fetch performance insights for a campaign."""
from __future__ import annotations

from domain.ports.outbound.ad_platform import IAdPlatform
from domain.ports.outbound.campaign_repository import ICampaignRepository

DEFAULT_FIELDS = ["impressions", "clicks", "spend", "ctr", "cpc", "reach"]


class GetInsightsUseCase:
    def __init__(self, ad_platform: IAdPlatform, campaign_repo: ICampaignRepository):
        self.ad_platform = ad_platform
        self.campaign_repo = campaign_repo

    async def execute(self, campaign_id: str, since: str, until: str) -> dict:
        campaign = await self.campaign_repo.get(campaign_id)
        if campaign is None:
            raise LookupError(f"Campaign {campaign_id} not found")
        if not campaign.meta_campaign_id:
            return {"data": []}
        return await self.ad_platform.get_insights(
            campaign.meta_campaign_id, DEFAULT_FIELDS, since, until
        )
